src/python/pants/engine/legacy/graph.py

```diff
@@ -296,6 +389,58 @@ class HydratedTargets(Collection.of(HydratedTarget)):
   """An intransitive set of HydratedTarget objects."""
 
 
+class OwnersRequest(datatype([
+  ('sources', tuple),
+  ('include_dependees', str),
+])):
+  """A request for the owners (and optionally, transitive dependees) of a set of file paths."""
+
+
+def find_owners(symbol_table, address_mapper, owners_request):
+  sources_set = set(owners_request.sources)
+  dirs_set = {dirname(source) for source in sources_set}
+  # Walk up the buildroot looking for targets that would conceivably claim changed sources.
+  candidate_specs = tuple(AscendantAddresses(directory=d) for d in dirs_set)
+
+  def owns_any_source(legacy_target):
+    """Given a `HydratedTarget` instance, check if it owns the given source file."""
+    target_kwargs = legacy_target.adaptor.kwargs()
+
+    # Handle `sources`-declaring targets.
+    # NB: Deleted files can only be matched against the 'filespec' (ie, `PathGlobs`) for a target,
+    # so we don't actually call `fileset.matches` here.
+    target_sources = target_kwargs.get('sources', None)
+    if target_sources and any_matches_filespec(sources_set, target_sources.filespec):
```

Comment:
I don't love that this means we have two implementations of glob matching (one in Python, one in Rust) which _do_ disagree for some excludes... (Python does pure regex, Rust does git-style ignore).

At a brief inspection, owners finding and the interpreter cache are the only things which use the python way - can we remove the python implementation (or only use it for the interpreter cache), and somehow delegate to the rust globs here? It shouldn't be a crazy thing to expose through the ffi...